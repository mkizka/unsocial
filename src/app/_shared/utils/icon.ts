import crypto from "crypto";

export const getIconHash = (icon: string) => {
  return crypto.createHash("md5").update(icon).digest("hex");
};

export const getIconPath = (
  iconHash: string | null | undefined,
  size: number,
) => {
  return `/icons/${iconHash ?? "default"}.webp?size=${size}`;
};
