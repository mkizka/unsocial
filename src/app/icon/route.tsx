import { notFound } from "next/navigation";
import type { NextRequest } from "next/server";
import { NextResponse } from "next/server";

import { env } from "@/utils/env";
import { prisma } from "@/utils/prisma";

export async function GET(request: NextRequest) {
  const preferredUsername =
    request.nextUrl.searchParams.get("preferredUsername");
  const host = request.nextUrl.searchParams.get("host");
  if (!preferredUsername || !host) {
    notFound();
  }
  const user = await prisma.user.findUnique({
    where: {
      preferredUsername_host: {
        preferredUsername,
        host,
      },
    },
  });
  if (!user) {
    notFound();
  }
  const url =
    user.icon ??
    `https://${env.HOST}/icon/edge?preferredUsername=${preferredUsername}`;
  const image = await fetch(url);
  return new NextResponse(await image.blob(), {
    headers: {
      "Content-Type": image.headers.get("Content-Type") ?? "image/png",
      // https://vercel.com/docs/concepts/functions/edge-functions/edge-caching#recommended-cache-control
      "Cache-Control": `max-age=0, s-maxage=${60 * 60 * 3}`,
    },
  });
}
