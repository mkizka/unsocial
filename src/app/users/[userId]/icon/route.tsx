import { notFound } from "next/navigation";
import { NextResponse } from "next/server";

import { env } from "@/utils/env";
import { prisma } from "@/utils/prisma";

export async function GET(
  _: Request,
  { params }: { params: { userId: string } },
) {
  const user = await prisma.user.findUnique({
    where: { id: params.userId },
  });
  if (!user) {
    notFound();
  }
  const url =
    user.icon ??
    `https://${env.HOST}/users/_/icon/edge?text=${user.preferredUsername}`;
  const image = await fetch(url);
  return new NextResponse(await image.blob(), {
    headers: {
      "Content-Type": image.headers.get("Content-Type") ?? "image/png",
      // https://vercel.com/docs/concepts/functions/edge-functions/edge-caching#recommended-cache-control
      "Cache-Control": `max-age=0, s-maxage=${60 * 60 * 3}`,
    },
  });
}
