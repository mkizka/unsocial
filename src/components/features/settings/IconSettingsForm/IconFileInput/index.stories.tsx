import type { Meta, StoryObj } from "@storybook/react";
import { http, HttpResponse } from "msw";

import { IconFileInput } from ".";
import icon from "./__fixtures__/icon.webp";
const meta: Meta<typeof IconFileInput> = {
  component: IconFileInput,
  args: {
    userId: "foo",
  },
  parameters: {
    msw: {
      handlers: [
        http.get("/users/foo/icon.webp?size=100", async (request) => {
          const file = await fetch(icon.src).then((res) => res.arrayBuffer());
          return HttpResponse.arrayBuffer(file, {
            headers: {
              "Content-Type": "image/webp",
            },
          });
        }),
      ],
    },
  },
};

export default meta;

type Story = StoryObj<typeof IconFileInput>;

export const Default: Story = {};
