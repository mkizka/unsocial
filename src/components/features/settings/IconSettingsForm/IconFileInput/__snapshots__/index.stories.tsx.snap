// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features/settings/IconSettingsForm/IconFileInput Default smoke-test 1`] = `
<div class="w-full rounded bg-primary-light p-4 shadow">
  <form class="space-y-1">
    <label class="block text-xl"
           for="icon"
    >
      アイコン
    </label>
    <input id="icon"
           accept="image/*"
           type="file"
    >
    <div class="flex">
      <div class="flex flex-col items-center space-y-1">
        <p>
          現在のアイコン
        </p>
        <img width="100"
             height="100"
             class="aspect-square object-cover"
             src="/users/foo/icon.webp?size=100"
             alt
        >
      </div>
      <div class="pt-16">
        <svg xmlns="http://www.w3.org/2000/svg"
             fill="none"
             viewbox="0 0 24 24"
             stroke-width="1.5"
             stroke="currentColor"
             aria-hidden="true"
             class="mx-4 h-6 w-6"
        >
          <path stroke-linecap="round"
                stroke-linejoin="round"
                d="M13.5 4.5L21 12m0 0l-7.5 7.5M21 12H3"
          >
          </path>
        </svg>
      </div>
      <div class="flex flex-col items-center space-y-1">
        <p>
          新しいアイコン
        </p>
        <div class="aspect-square w-[100px] bg-gray">
        </div>
      </div>
    </div>
  </form>
</div>
`;
